File: 26. HighestPower_MSB.py
# Topic: Highest power of 2 <= N and base-2 log (position of MSB)

def highest_power_of_2_le(n):
    if n < 1: return 0
    p = 1
    while p <= n: p <<= 1
    return p >> 1

def msb_pos(n):
    if n <= 0: return -1
    return n.bit_length() - 1

if __name__ == '__main__':
    print('Highest pow2 <=20:', highest_power_of_2_le(20))
    print('MSB pos 16:', msb_pos(16))

# Time Complexity: O(1)
# Space Complexity: O(1)

File: 26. HighestPower_MSB.cpp
// Topic: Highest power of 2 <= N and base-2 log (position of MSB)
#include <bits/stdc++.h>
using namespace std;

int highestPowerOf2LE(int n)
{
    if (n < 1)
        return 0;
    int p = 1;
    while (p <= n) p <<= 1;
    return p >> 1;
}
int msbPos(int n)
{
    if (n <= 0)
        return -1;
    return 31 - __builtin_clz(n);
}

int main()
{
    cout << "Highest pow2 <=20: " << highestPowerOf2LE(20) << '\n';
    cout << "MSB pos 16: " << msbPos(16) << '\n';
    return 0;
}

// Time Complexity: O(1)
// Space Complexity: O(1)
